homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -0,0 +1,306 @@
+"""Support for hunter douglas shades."""
+import asyncio
+import logging
+
+from aiopvapi.helpers.constants import ATTR_POSITION1, ATTR_POSITION_DATA
+from aiopvapi.resources.shade import (
+    ATTR_POSKIND1,
+    ATTR_TYPE,
+    MAX_POSITION,
+    MIN_POSITION,
+    factory as PvShade,
+)
+import async_timeout
+
+from homeassistant.components.cover import (
+    ATTR_POSITION,
+    DEVICE_CLASS_SHADE,
+    SUPPORT_CLOSE,
+    SUPPORT_OPEN,
+    SUPPORT_SET_POSITION,
+    SUPPORT_STOP,
+    CoverEntity,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_call_later
+
+from .const import (
+    COORDINATOR,
+    DEVICE_INFO,
+    DEVICE_MODEL,
+    DEVICE_SERIAL_NUMBER,
+    DOMAIN,
+    FIRMWARE_BUILD,
+    FIRMWARE_IN_SHADE,
+    FIRMWARE_REVISION,
+    FIRMWARE_SUB_REVISION,
+    MANUFACTURER,
+    PV_API,
+    PV_ROOM_DATA,
+    PV_SHADE_DATA,
+    ROOM_ID_IN_SHADE,
+    ROOM_NAME_UNICODE,
+    SHADE_RESPONSE,
+    STATE_ATTRIBUTE_ROOM_NAME,
+)
+from .entity import HDEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+# Estimated time it takes to complete a transition
+# from one state to another
+TRANSITION_COMPLETE_DURATION = 30
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the hunter douglas shades."""
+
+    pv_data = hass.data[DOMAIN][entry.entry_id]
+    room_data = pv_data[PV_ROOM_DATA]
+    shade_data = pv_data[PV_SHADE_DATA]
+    pv_request = pv_data[PV_API]
+    coordinator = pv_data[COORDINATOR]
+    device_info = pv_data[DEVICE_INFO]
+
+    entities = []
+    for raw_shade in shade_data.values():
+        # The shade may be out of sync with the hub
+        # so we force a refresh when we add it if
+        # possible
+        shade = PvShade(raw_shade, pv_request)
+        name_before_refresh = shade.name
+        try:
+            async with async_timeout.timeout(1):
+                await shade.refresh()
```

Comment:
Consider creating a function and run all shades in parallel.

Reply:
I tried this approach first but managed to get the hub to stop responding